# kb_gtdbtk

Build status (master):
[![Build Status](https://travis-ci.org/kbaseapps/kb_gtdbtk.svg?branch=master)](https://travis-ci.org/kbaseapps/kb_gtdbtk)
[![Coverage Status](https://coveralls.io/repos/github/kbaseapps/kb_gtdbtk/badge.svg?branch=master)](https://coveralls.io/github/kbaseapps/kb_gtdbtk?branch=master)

This is a [KBase](https://kbase.us) module generated by the [KBase Software Development Kit (SDK)](https://github.com/kbase/kb_sdk).

You will need to have the SDK installed to use this module. [Learn more about the SDK and how to use it](https://kbase.github.io/kb_sdk_docs/).

You can also learn more about the apps implemented in this module from its [catalog page](https://narrative.kbase.us/#catalog/modules/kb_gtdbtk) or its [spec file]($module_name.spec).

# GTDB-Tk

This KBase apps wraps the GTDB-Tk commandline tool created by Donavan Parks, Pierre-Alain Chaumeil, and Phil Hugenholtz of the [Austrailian Centre for Ecogenomics](https://gtdb.ecogenomics.org)

[GTDB-Tk](https://github.com/Ecogenomics/GtdbTk) is a software toolkit for assigning objective taxonomic classifications to bacterial and archaeal genomes. It is designed to work with recent advances that allow hundreds or thousands of metagenome-assembled genomes (MAGs) to be obtained directly from environmental samples. It can also be applied to isolate and single-cell genomes. The GTDB-Tk is open source and released under the GNU General Public License (Version 3).

## References

* Chaumeil PA, Mussig AJ, Hugenholtz P, Parks DH. 2019. [GTDB-Tk: A toolkit to classify genomes with the Genome Taxonomy Database](https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btz848/5626182). <i>Bioinformatics</i>, btz848.

* Parks DH, et al. 2018. [A standardized bacterial taxonomy based on genome phylogeny substantially revises the tree of life](https://www.nature.com/articles/nbt.4229). <i>Nat. Biotechnol.</i>, http://dx.doi.org/10.1038/nbt.4229

 We also strongly encourage you to cite the following 3rd party dependencies:

* Matsen FA, Kodner RB, Armbrust EV. 2010. [pplacer: linear time maximum-likelihood and Bayesian phylogenetic placement of sequences onto a fixed reference tree](https://www.ncbi.nlm.nih.gov/pubmed/21034504). <i>BMC Bioinformatics</i>, 11:538.
* Jain C, et al. 2017. [High-throughput ANI Analysis of 90K Prokaryotic Genomes Reveals Clear Species Boundaries](https://www.biorxiv.org/content/early/2017/11/27/225342). <i>bioRxiv</i>, https://doi.org/10.1101/225342.
* Hyatt D, et al. 2010. [Prodigal: prokaryotic gene recognition and translation initiation site identification](https://www.ncbi.nlm.nih.gov/pubmed/20211023). <i>BMC Bioinformatics</i>, 11:119. doi: 10.1186/1471-2105-11-119.
* Price MN, Dehal PS, Arkin AP. [FastTree 2 - Approximately Maximum-Likelihood Trees for Large Alignments](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2835736/). <i>PLoS One</i>, 5, e9490.
* Eddy SR. 2011. [Accelerated profile HMM searches](https://www.ncbi.nlm.nih.gov/pubmed/22039361). <i>PLOS Comp. Biol.</i>, 7:e1002195.

# Setup and test

## Test modes

GTDB-tk v1.7.0 takes a minimum of [212 GB of memory and 51 GB of disk to run](https://github.com/Ecogenomics/GTDBTk#hardware-requirements).
Therefore it is generally not feasible to test in a CI environment (e.g. Travis-CI) or on a
developer's personal computer.

As such there are two test modes supported - one that tests a subset of the code but requires
minimal resources, and a full test suite.

The former is suitable for rapid development, but the full suite should always be run before
releasing new code to production.

### Mocked test mode

In this mode, the GTDB-tk application and any calls to KBase services or the SDK callback
server are mocked and only unit tests are run. This mode is run in Travis-CI and the coverage
reported by Coveralls is based on this suite of tests.

To run mocked mode, from the module root directory:

```bash
pipenv shell
make test-sdkless
```

If pipenv is not installed, do `pip install pipenv`.

Running tests in this mode requires no reference data, very little memory, and takes a few seconds.
However, it **does not** test the code in the Impl file.

### Full test mode

In this mode, the GTDB-tk application is run and any KBase services in the specified environment
are contacted as in a normal SDK test run. As this requires a token, this test mode cannot be
run in Travis-CI for Github PRs.

To run full mode, first add your KBase developer token to `test_local/test.cfg`. Then from the
module root directory run:

```bash
make  # may be omitted after the first run
kb-sdk test
```

Running tests in this mode requires 51 GB of reference data, 212 GB of memory, and takes on the
order of an hour. It runs all the tests, which includes an integration test with a KBase Assembly
object. More tests should be added in the future. For KBase developers, the `dev1` machine in Chicago
is a suitable place to run the full test suite and the reference data is already available in
the `/kb/data/kb_gtdbtk` directory.

## Testing TODO

* Add integration tests for the other processed types.
  * Experiment with assembly sizes in order to have GTDB-tk recognize genes. The current
    test data is too small for this.
* Add copy of a copy integration tests (requires at least one more token).
* Add failing integration test cases

# Installation from another module

To use this code in another SDK module, call `kb-sdk install kb_gtdbtk` in the other module's root directory.

# Help

You may find the answers to your questions in our [FAQ](https://kbase.github.io/kb_sdk_docs/references/questions_and_answers.html) or [Troubleshooting Guide](https://kbase.github.io/kb_sdk_docs/references/troubleshooting.html).
